### ISSUE TITLE
Patroni v3, fails to get list of machines

### ISSUE THREAD

**Issue by pgodfrin-nov** [zalando/patroni#1939]
**Describe the bug**
Starting Patroni which fails to get list of machines:
```
postgres@TSTSHD01:/postgres/admin/patroni=> patroni /postgres/admin/patroni/tstshd01-patroni.yaml 
2021-05-21 16:03:24,204 ERROR: Failed to get list of machines from http://172.31.77.187:2379/v3: <Unknown error: '404 page not found', code: 2>
2021-05-21 16:03:24,205 ERROR: Failed to get list of machines from http://172.31.68.147:2379/v3: <Unknown error: '404 page not found', code: 2>
2021-05-21 16:03:24,206 ERROR: Failed to get list of machines from http://172.31.69.227:2379/v3: <Unknown error: '404 page not found', code: 2>
2021-05-21 16:03:24,206 INFO: waiting on etcd
2021-05-21 16:03:29,214 ERROR: Failed to get list of machines from http://172.31.77.187:2379/v3: <Unknown error: '404 page not found', code: 2>
2021-05-21 16:03:29,215 ERROR: Failed to get list of machines from http://172.31.68.147:2379/v3: <Unknown error: '404 page not found', code: 2>
2021-05-21 16:03:29,216 ERROR: Failed to get list of machines from http://172.31.69.227:2379/v3: <Unknown error: '404 page not found', code: 2>
2021-05-21 16:03:29,216 INFO: waiting on etcd
20
```

**To Reproduce**
Configured:
```
```
# etcd v3
etcd3:
  #Provide host to do the initial discovery of the cluster topology:
  # host: 127.0.0.1:2379
  #Or use "hosts" to provide multiple endpoints
  #Could be a comma separated string:
  #hosts: host1:port1,host2:port2
  #or an actual yaml list:
  hosts:
  - 172.31.77.187:2379
  - 172.31.68.147:2379
  - 172.31.69.227:2379
  # - prdetcdwit02:2379
  #- host1:port1
  #- host2:port2
  # url:  ??? "url for the etcd" have no idea what that means
  #Once discovery is complete Patroni will use the list of advertised clientURLs
  #It is possible to change this behavior through by setting:
  #use_proxies: true
```

**Expected behavior**
successful response, 404
**Screenshots**
If applicable, add screenshots to help explain your problem.

**Environment**
- Patroni version:
```
postgres@TSTSHD01:/postgres/admin/patroni=> patroni --version
patroni 2.0.2
```
- PostgreSQL version:
```
postgres@TSTSHD01:/postgres/admin/patroni=> psql
psql (13.3 (Ubuntu 13.3-1.pgdg18.04+1))
Type "help" for help.
postgres=# 
```
- DCS (and its version):
```
postgres@TSTSHD01:/postgres/admin/patroni=> eew version  
etcdctl version: 3.4.16
API version: 3.4
```
**Patroni configuration file**
```
Please copy&paste your Patroni configuration file here
```
postgres@TSTSHD01:/postgres/admin/patroni=> cat tstshd01-patroni.yaml 
# test shard 01
# tstshd01
# more details: https://patroni.readthedocs.io/en/latest/SETTINGS.html#settings
#
scope: etcd-tstshd01
#namespace: /tstpat01
name: etcd-tstshd01

restapi:
  listen: 127.0.0.1:8008
  connect_address: 127.0.0.1:8008
#  certfile: /etc/ssl/certs/ssl-cert-snakeoil.pem
#  keyfile: /etc/ssl/private/ssl-cert-snakeoil.key
#  authentication:
#    username: username
#    password: password

# ctl:
#   insecure: false # Allow connections to SSL sites without certs
#   certfile: /etc/ssl/certs/ssl-cert-snakeoil.pem
#   cacert: /etc/ssl/certs/ssl-cacert-snakeoil.pem


# peg 05/17/21
###PEG###172.31.77.193   tsthpx01
###PEG###172.31.68.147   tstshd01
###PEG###172.31.69.227   tstshd02
###PEG###172.31.77.187   tstewt01
# etcd v3
etcd3:
  #Provide host to do the initial discovery of the cluster topology:
  # host: 127.0.0.1:2379
  #Or use "hosts" to provide multiple endpoints
  #Could be a comma separated string:
  #hosts: host1:port1,host2:port2
  #or an actual yaml list:
  hosts:
  - 172.31.77.187:2379
  - 172.31.68.147:2379
  - 172.31.69.227:2379
  # - prdetcdwit02:2379
  #- host1:port1
  #- host2:port2
  # url:  ??? "url for the etcd" have no idea what that means
  #Once discovery is complete Patroni will use the list of advertised clientURLs
  #It is possible to change this behavior through by setting:
  #use_proxies: true

# etcd v2
# etcd:
  # # Provide host to do the initial discovery of the cluster topology:
  # host: 127.0.0.1:2379
  # # Or use "hosts" to provide multiple endpoints
  # # Could be a comma separated string:
  # # hosts: host1:port1,host2:port2
  # # or an actual yaml list:
  # # hosts:
  # # - host1:port1
  # # - host2:port2
  # # Once discovery is complete Patroni will use the list of advertised clientURLs
  # # It is possible to change this behavior through by setting:
  # # use_proxies: true

#raft:
#  data_dir: .
#  self_addr: 127.0.0.1:2222
#  partner_addrs:
#  - 127.0.0.1:2223
#  - 127.0.0.1:2224

bootstrap:
  # this section will be written into Etcd:/<namespace>/<scope>/config after initializing new cluster
  # and all other cluster members will use it as a `global configuration`
  dcs:
    ttl: 30
    loop_wait: 10
    retry_timeout: 10
    # the maximum bytes a follower may lag to be able to participate in leader election.
    # compare this to historical replication times
    maximum_lag_on_failover: 1048576
    # the amount of time a master is allowed to recover from failures before failover 
    # is triggered (in seconds). Default is 300 seconds
#    master_start_timeout: 300
    # The number of seconds Patroni is allowed to wait when stopping Postgres and 
    # effective only when synchronous_mode is enabled. 
#    master_stop_timeout:
#    synchronous_mode: false
    #standby_cluster:
      #host: 127.0.0.1
      #port: 1111
      #primary_slot_name: patroni
    postgresql:
      use_pg_rewind: true
      # not needed for streaming
      use_slots: false
      parameters:
#        wal_level: hot_standby
#        hot_standby: "on"
#        max_connections: 100
#        max_worker_processes: 8
#        wal_keep_segments: 8
#        max_wal_senders: 10
#        max_replication_slots: 10
#        max_prepared_transactions: 0
#        max_locks_per_transaction: 64
#        wal_log_hints: "on"
#        track_commit_timestamp: "off"
#        archive_mode: "on"
#        archive_timeout: 1800s
#        archive_command: mkdir -p ../wal_archive && test ! -f ../wal_archive/%f && cp %p ../wal_archive/%f
#      recovery_conf:
#        restore_command: cp ../wal_archive/%f %p

  # pgodfrin 05-11 bootstrapping is presumably the process when
  # patroni is first started up. It's questionable if a database
  # will need to be intialized. every time patroni is bootstrapped
  # some desired options for 'initdb'
###PEG###  I never want patroni to intialize the DB
###PEG###  initdb:  # Note: It needs to be a list (some options need values, others are switches)
###PEG###  - encoding: UTF8
###PEG###  - data-checksums

  pg_hba:  # Add following lines to pg_hba.conf after running 'initdb'
  # For kerberos gss based connectivity (discard @.*$)
  #- host replication replicator 127.0.0.1/32 gss include_realm=0
  #- host all all 0.0.0.0/0 gss include_realm=0
  - host replication replicator 127.0.0.1/32 md5
  - host all all 0.0.0.0/0 md5
#  - hostssl all all 0.0.0.0/0 md5

  # Additional script to be launched after initial cluster creation (will be passed the connection URL as parameter)
# post_init: /usr/local/bin/setup_cluster.sh

  # Some additional users users which needs to be created after initializing new cluster
  # users:
    # admin:
      # password: admin
      # options:
        # - createrole
        # - createdb

postgresql:
  listen: 127.0.0.1:5432
  connect_address: 127.0.0.1:5432
  data_dir: /postgres/shd01/data
#  bin_dir:
#  config_dir:
  pgpass: /home/postgres/.patroni/.pgpass
  authentication:
    replication:
      username: standby
      # password: rep-pass
    superuser:
      username: postgres
      # password: zalando
    rewind:  # Has no effect on postgres 10 and lower
      username: postgres
      # password: rewind_password
  # Server side kerberos spn
#  krbsrvname: postgres
  parameters:
    # Fully qualified kerberos ticket file for the running user
    # same as KRB5CCNAME used by the GSS
#   krb_server_keyfile: /var/spool/keytabs/postgres
    unix_socket_directories: '/var/run/postgresql'
  # Additional fencing script executed after acquiring the leader lock but before promoting the replica
  #pre_promote: /path/to/pre_promote.sh

#watchdog:
#  mode: automatic # Allowed values: off, automatic, required
#  device: /dev/watchdog
#  safety_margin: 5

tags:
    nofailover: false
    noloadbalance: false
    clonefrom: false
    nosync: false
```
**patronictl show-config**
```
postgres@TSTSHD01:/postgres/admin/patroni=> patronictl show-config
Error: 'Can not find suitable configuration of distributed configuration store\nAvailable implementations: etcd, etcd3, kubernetes'```

**Have you checked Patroni logs?**
postgres@TSTSHD01:/postgres/admin/patroni=> patroni /postgres/admin/patroni/tstshd01-patroni.yaml 
2021-05-21 17:27:52,614 ERROR: Failed to get list of machines from http://172.31.77.187:2379/v3: <Unknown error: '404 page not found', code: 2>
2021-05-21 17:27:52,615 ERROR: Failed to get list of machines from http://172.31.68.147:2379/v3: <Unknown error: '404 page not found', code: 2>
2021-05-21 17:27:52,616 ERROR: Failed to get list of machines from http://172.31.69.227:2379/v3: <Unknown error: '404 page not found', code: 2>
2021-05-21 17:27:52,616 INFO: waiting on etcd
2021-05-21 17:27:57,624 ERROR: Failed to get list of machines from http://172.31.77.187:2379/v3: <Unknown error: '404 page not found', code: 2>
2021-05-21 17:27:57,625 ERROR: Failed to get list of machines from http://172.31.68.147:2379/v3: <Unknown error: '404 page not found', code: 2>
2021-05-21 17:27:57,630 ERROR: Failed to get list of machines from http://172.31.69.227:2379/v3: <Unknown error: '404 page not found', code: 2>
2021-05-21 17:27:57,630 INFO: waiting on etcd
2021-05-21 17:28:02,634 ERROR: Failed to get list of machines from http://172.31.77.187:2379/v3: <Unknown error: '404 page not found', code: 2>
2021-05-21 17:28:02,635 ERROR: Failed to get list of machines from http://172.31.68.147:2379/v3: <Unknown error: '404 page not found', code: 2>
2021-05-21 17:28:02,636 ERROR: Failed to get list of machines from http://172.31.69.227:2379/v3: <Unknown error: '404 page not found', code: 2>
2021-05-21 17:28:02,637 INFO: waiting on etcd
^CTraceback (most recent call last):
  File "/usr/lib/python3/dist-packages/patroni/dcs/etcd.py", line 382, in _refresh_machines_cache
    self._machines_cache = self.machines
  File "/usr/lib/python3/dist-packages/patroni/dcs/etcd.py", line 197, in machines
    raise etcd.EtcdConnectionFailed('No more machines in the cluster')
etcd.EtcdConnectionFailed: No more machines in the cluster

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/patroni/dcs/etcd.py", line 535, in get_etcd_client
    client = client_cls(config, dns_resolver)
  File "/usr/lib/python3/dist-packages/patroni/dcs/etcd3.py", line 513, in __init__
    super(PatroniEtcd3Client, self).__init__(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/patroni/dcs/etcd3.py", line 175, in __init__
    super(Etcd3Client, self).__init__(config, dns_resolver, cache_ttl)
  File "/usr/lib/python3/dist-packages/patroni/dcs/etcd.py", line 101, in __init__
    self._load_machines_cache()
  File "/usr/lib/python3/dist-packages/patroni/dcs/etcd.py", line 372, in _load_machines_cache
    self._refresh_machines_cache(True)
  File "/usr/lib/python3/dist-packages/patroni/dcs/etcd.py", line 385, in _refresh_machines_cache
    raise etcd.EtcdException("Could not get the list of servers, "
etcd.EtcdException: Could not get the list of servers, maybe you provided the wrong host(s) to connect to?

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/bin/patroni", line 11, in <module>
    load_entry_point('patroni==2.0.2', 'console_scripts', 'patroni')()
  File "/usr/lib/python3/dist-packages/patroni/__init__.py", line 170, in main
    return patroni_main()
  File "/usr/lib/python3/dist-packages/patroni/__init__.py", line 138, in patroni_main
    abstract_main(Patroni, schema)
  File "/usr/lib/python3/dist-packages/patroni/daemon.py", line 98, in abstract_main
    controller = cls(config)
  File "/usr/lib/python3/dist-packages/patroni/__init__.py", line 29, in __init__
    self.dcs = get_dcs(self.config)
  File "/usr/lib/python3/dist-packages/patroni/dcs/__init__.py", line 92, in get_dcs
    return item(config[name])
  File "/usr/lib/python3/dist-packages/patroni/dcs/etcd3.py", line 578, in __init__
    super(Etcd3, self).__init__(config, PatroniEtcd3Client, (DeadlineExceeded, Unavailable, FailedPrecondition))
  File "/usr/lib/python3/dist-packages/patroni/dcs/etcd.py", line 432, in __init__
    self._client = self.get_etcd_client(config, client_cls)
  File "/usr/lib/python3/dist-packages/patroni/dcs/etcd.py", line 540, in get_etcd_client
    time.sleep(5)
KeyboardInterrupt
**Have you checked PostgreSQL logs?**
Please provide a snippet here

**Have you tried to use GitHub issue search?**
Maybe there is already a similar issue solved.


**Additional context**
Add any other context about the problem here.

**Comment by CyberDem0n** [comment]
> postgres@TSTSHD01:/postgres/admin/patroni=> psql
psql (13.3 (Ubuntu 13.3-1.pgdg18.04+1))
Type "help" for help.
postgres=#

Why the postgres is up and running without being supervised by Patroni?

> postgres@TSTSHD01:/postgres/admin/patroni=> eew version
etcdctl version: 3.4.16
API version: 3.4

The questionnaire explicitly asking about the version of Etcd, not etcdctl.
I have no idea what `eew version` is doing.
You can get the version of etcd by running `curl $ETCD_IP:2379/version`.

> 2021-05-21 17:27:52,614 ERROR: Failed to get list of machines from http://172.31.77.187:2379/v3: <Unknown error: '404 page not found', code: 2>

That clearly means that whatever is running on `172.31.77.187:2379` doesn't speak Etcd v3 protocol.
If there is really Etcd v3 running, it could respond like:
```
$ curl -XPOST localhost:2379/v3/cluster/member/list
{"header":{"cluster_id":"9090345565864637102","member_id":"14855829450254237642","raft_term":"3"},
  "members":[{"ID":"14855829450254237642","name":"default",
  "peerURLs":["http://localhost:2380","http://localhost:7001"],
  "clientURLs":["http://localhost:2379"]}]}
```

**Comment by pgodfrin-nov** [comment]
I tried it with both v2 and v3:
```2021-05-21 18:56:49,812 ERROR: Failed to get list of machines from http://172.31.77.187:2379/v3: <Unknown error: '404 page not found', code: 2>
2021-05-21 18:56:49,813 DEBUG: Converted retries value: 0 -> Retry(total=0, connect=None, read=None, redirect=0, status=None)
2021-05-21 18:56:49,813 DEBUG: Starting new HTTP connection (1): 172.31.68.147
2021-05-21 18:56:49,814 DEBUG: http://172.31.68.147:2379 "POST /v3/cluster/member/list HTTP/1.1" 404 19
2021-05-21 18:56:49,814 ERROR: Failed to get list of machines from http://172.31.68.147:2379/v3: <Unknown error: '404 page not found', code: 2>
2021-05-21 18:56:49,814 DEBUG: Converted retries value: 0 -> Retry(total=0, connect=None, read=None, redirect=0, status=None)
2021-05-21 18:56:49,814 DEBUG: Starting new HTTP connection (1): 172.31.69.227
2021-05-21 18:56:49,815 DEBUG: http://172.31.69.227:2379 "POST /v3/cluster/member/list HTTP/1.1" 404 19
2021-05-21 18:56:49,816 ERROR: Failed to get list of machines from http://172.31.69.227:2379/v3: <Unknown error: '404 page not found', code: 2>
2021-05-21 18:56:49,816 INFO: waiting on etcd
2021-05-21 19:26:23,447 DEBUG: New etcd client created for http://127.0.0.1:4001
2021-05-21 19:26:23,448 DEBUG: Converted retries value: 0 -> Retry(total=0, connect=None, read=None, redirect=0, status=None)
2021-05-21 19:26:23,448 DEBUG: Starting new HTTP connection (1): 172.31.77.187
2021-05-21 19:26:23,449 DEBUG: http://172.31.77.187:2379 "GET /v2/machines HTTP/1.1" 404 19
2021-05-21 19:26:23,449 ERROR: Failed to get list of machines from http://172.31.77.187:2379/v2: EtcdException('Bad response : 404 page not found\n',)
2021-05-21 19:26:23,450 DEBUG: Converted retries value: 0 -> Retry(total=0, connect=None, read=None, redirect=0, status=None)
2021-05-21 19:26:23,450 DEBUG: Starting new HTTP connection (1): 172.31.68.147
2021-05-21 19:26:23,450 DEBUG: http://172.31.68.147:2379 "GET /v2/machines HTTP/1.1" 404 19
2021-05-21 19:26:23,451 ERROR: Failed to get list of machines from http://172.31.68.147:2379/v2: EtcdException('Bad response : 404 page not found\n',)
2021-05-21 19:26:23,451 DEBUG: Converted retries value: 0 -> Retry(total=0, connect=None, read=None, redirect=0, status=None)
2021-05-21 19:26:23,451 DEBUG: Starting new HTTP connection (1): 172.31.69.227
2021-05-21 19:26:23,452 DEBUG: http://172.31.69.227:2379 "GET /v2/machines HTTP/1.1" 404 19
2021-05-21 19:26:23,452 ERROR: Failed to get list of machines from http://172.31.69.227:2379/v2: EtcdException('Bad response : 404 page not found\n',)
2021-05-21 19:26:23,452 INFO: waiting on etcd
```

**Comment by CyberDem0n** [comment]
>I tried it with both v2 and v3:

You don't read what I tell you, do you?

**Comment by pgodfrin-nov** [comment]
> postgres@TSTSHD01:/postgres/admin/patroni=> psql
psql (13.3 (Ubuntu 13.3-1.pgdg18.04+1))
Type "help" for help.
postgres=#

Why the postgres is up and running without being supervised by Patroni?

**Good point, I shutdown postgres and got the same error.** 
postgres@TSTSHD01:/postgres/admin/patroni=> eew version
etcdctl version: 3.4.16
API version: 3.4

The questionnaire explicitly asking about the version of Etcd, not etcdctl.
I have no idea what eew version is doing.
You can get the version of etcd by running curl $ETCD_IP:2379/version.
**sorry about that eew is just an alias**
```
postgres@TSTSHD01:/postgres/admin/patroni/logs=> curl 172.31.69.227:2379/version
{"etcdserver":"3.4.16","etcdcluster":"3.4.0"}postgres@TSTSHD01:/postgres/admin/patroni/logs=> 
```




2021-05-21 17:27:52,614 ERROR: Failed to get list of machines from http://172.31.77.187:2379/v3: <Unknown error: '404 page not found', code: 2>

That clearly means that whatever is running on 172.31.77.187:2379 doesn't speak Etcd v3 protocol.
If there is really Etcd v3 running, it could respond like:

$ curl -XPOST localhost:2379/v3/cluster/member/list
{"header":{"cluster_id":"9090345565864637102","member_id":"14855829450254237642","raft_term":"3"},
  "members":[{"ID":"14855829450254237642","name":"default",
  "peerURLs":["http://localhost:2380","http://localhost:7001"],
  "clientURLs":["http://localhost:2379"]}]}
**OK, thanks I'm really a DBA, so I'm just learning about this stuff**
**Definitely doesn't like the loopback/localhost**
```
postgres@TSTSHD01:/postgres/admin/patroni/logs=> curl -XPOST localhost:2379/v3/cluster/member/list
curl: (7) Failed to connect to localhost port 2379: Connection refused
```
**I added local host to the peer listeners, now the XPOST get's a 404**
```
postgres@TSTSHD01:/postgres/admin/etcd=> curl -XPOST localhost:2379/v3/cluster/member/list
404 page not found
```
**Looks like etc is running:**
```
postgres@TSTSHD01:/postgres/admin/etcd=> ps -ef | grep etcd | grep -v grep
postgres 21472 18904  1 19:39 pts/2    00:00:02 etcd --config-file /postgres/admin/etcd/tstshd01-etcd.yaml

postgres@TSTSHD01:/postgres/admin/etcd=> eee endpoint health
+--------------------+--------+------------+-------+
|      ENDPOINT      | HEALTH |    TOOK    | ERROR |
+--------------------+--------+------------+-------+
| 172.31.68.147:2379 |   true | 2.989743ms |       |
| 172.31.77.187:2379 |   true | 1.900477ms |       |
| 172.31.69.227:2379 |   true | 2.590568ms |       |
+--------------------+--------+------------+-------+
postgres@TSTSHD01:/postgres/admin/etcd=> eee endpoint status
+--------------------+------------------+---------+---------+-----------+------------+-----------+------------+--------------------+--------+
|      ENDPOINT      |        ID        | VERSION | DB SIZE | IS LEADER | IS LEARNER | RAFT TERM | RAFT INDEX | RAFT APPLIED INDEX | ERRORS |
+--------------------+------------------+---------+---------+-----------+------------+-----------+------------+--------------------+--------+
| 172.31.68.147:2379 | 87cff2241217dcd4 |  3.4.16 |   43 MB |     false |      false |       233 |      19092 |              19092 |        |
| 172.31.69.227:2379 | 43af966b59a8355f |  3.4.16 |   43 MB |     false |      false |       233 |      19092 |              19092 |        |
| 172.31.77.187:2379 | ec7afc04fb7579da |  3.4.16 |   43 MB |      true |      false |       233 |      19092 |              19092 |        |
+--------------------+------------------+---------+---------+-----------+------------+-----------+------------+--------------------+--------+
```

**Comment by CyberDem0n** [comment]
I am not sure how you did it, but the JSON gRPC gateway in etcd seems to be disabled.
You have to enable it back.

**Comment by pgodfrin-nov** [comment]
roger that, on it now, thanks!

**Comment by pgodfrin-nov** [comment]
No joy:

```
postgres@TSTSHD01:/postgres/admin/patroni/logs=> tail -f patroni.log  
2021-05-21 21:15:47,815 ERROR: Failed to get list of machines from http://172.31.77.187:2379/v3: <Unknown error: '404 page not found', code: 2>
2021-05-21 21:15:47,815 DEBUG: Converted retries value: 0 -> Retry(total=0, connect=None, read=None, redirect=0, status=None)
2021-05-21 21:15:47,816 DEBUG: Starting new HTTP connection (1): 172.31.68.147
2021-05-21 21:15:47,816 DEBUG: http://172.31.68.147:2379 "POST /v3/cluster/member/list HTTP/1.1" 404 19
2021-05-21 21:15:47,816 ERROR: Failed to get list of machines from http://172.31.68.147:2379/v3: <Unknown error: '404 page not found', code: 2>
2021-05-21 21:15:47,817 DEBUG: Converted retries value: 0 -> Retry(total=0, connect=None, read=None, redirect=0, status=None)
2021-05-21 21:15:47,817 DEBUG: Starting new HTTP connection (1): 172.31.69.227
2021-05-21 21:15:47,818 DEBUG: http://172.31.69.227:2379 "POST /v3/cluster/member/list HTTP/1.1" 404 19
2021-05-21 21:15:47,818 ERROR: Failed to get list of machines from http://172.31.69.227:2379/v3: <Unknown error: '404 page not found', code: 2>
2021-05-21 21:15:47,818 INFO: waiting on etcd
2021-05-21 21:15:52,823 DEBUG: New etcd client created for http://127.0.0.1:4001
2021-05-21 21:15:52,824 DEBUG: Converted retries value: 0 -> Retry(total=0, connect=None, read=None, redirect=0, status=None)
2021-05-21 21:15:52,824 DEBUG: Starting new HTTP connection (1): 172.31.77.187
2021-05-21 21:15:52,825 DEBUG: http://172.31.77.187:2379 "GET /version HTTP/1.1" 200 45
2021-05-21 21:15:52,826 DEBUG: Converted retries value: 0 -> Retry(total=0, connect=None, read=None, redirect=0, status=None)
2021-05-21 21:15:52,826 DEBUG: http://172.31.77.187:2379 "POST /v3/cluster/member/list HTTP/1.1" 404 19
2021-05-21 21:15:52,827 ERROR: Failed to get list of machines from http://172.31.77.187:2379/v3: <Unknown error: '404 page not found', code: 2>
2021-05-21 21:15:52,827 DEBUG: Converted retries value: 0 -> Retry(total=0, connect=None, read=None, redirect=0, status=None)
2021-05-21 21:15:52,827 DEBUG: Starting new HTTP connection (1): 172.31.68.147
2021-05-21 21:15:52,828 DEBUG: http://172.31.68.147:2379 "POST /v3/cluster/member/list HTTP/1.1" 404 19
2021-05-21 21:15:52,828 ERROR: Failed to get list of machines from http://172.31.68.147:2379/v3: <Unknown error: '404 page not found', code: 2>
2021-05-21 21:15:52,828 DEBUG: Converted retries value: 0 -> Retry(total=0, connect=None, read=None, redirect=0, status=None)
2021-05-21 21:15:52,828 DEBUG: Starting new HTTP connection (1): 172.31.69.227
2021-05-21 21:15:52,829 DEBUG: http://172.31.69.227:2379 "POST /v3/cluster/member/list HTTP/1.1" 404 19
2021-05-21 21:15:52,830 ERROR: Failed to get list of machines from http://172.31.69.227:2379/v3: <Unknown error: '404 page not found', code: 2>
2021-05-21 21:15:52,830 INFO: waiting on etcd
^C
postgres@TSTSHD01:/postgres/admin/patroni/logs=> ps -ef | grep post
root     14901  6672  0 May20 pts/0    00:00:00 su - postgres
postgres 14902 14901  0 May20 pts/0    00:00:01 -su
root     18903 18888  0 14:56 pts/2    00:00:00 su - postgres
postgres 18904 18903  0 14:56 pts/2    00:00:00 -su
postgres 21869 14902  0 21:02 pts/0    00:00:07 etcd --config-file /postgres/admin/etcd/tstshd01-etcd.yaml
postgres 21917 18904  0 21:05 pts/2    00:00:00 etcd grpc-proxy start --endpoints=172.31.68.147:2379,172.31.69.227:2379,172.31.77.187:2379 --listen-addr=127.0.0.1:2379
postgres 22064 14902  1 21:15 pts/0    00:00:00 /usr/bin/python3 /usr/bin/patroni /postgres/admin/patroni/tstshd01-patroni.yaml
postgres 22085 14902  0 21:16 pts/0    00:00:00 ps -ef
postgres 22086 14902  0 21:16 pts/0    00:00:00 grep --color=auto post
```

I can RW manually:
```
postgres@TSTSHD01:/postgres/admin/etcd=> ETCDCTL_API=3 etcdctl --endpoints=127.0.0.1:2379 put foo bar
OK
postgres@TSTSHD01:/postgres/admin/etcd=> ETCDCTL_API=3 etcdctl --endpoints=127.0.0.1:2379 get foo
foo
bar
```
still no joy:
```
postgres@TSTSHD01:/postgres/admin/patroni/logs=> curl -XPOST localhost:2379/v3/cluster/member/list
404 page not found
```

**Comment by CyberDem0n** [comment]
You don't need to run the gRPC gateway separately. It must be enabled on all nodes of etcd cluster.
See https://github.com/etcd-io/etcd/issues/12093

**Comment by pgodfrin-nov** [comment]
So far that seems to have worked. Is etcd-io/etcd#12093 in the docs?

**Comment by CyberDem0n** [comment]
Which docs? You are the first one who encountered this configuration quirk.

**Comment by pgodfrin-nov** [comment]
OK - thanks CyberDem0n, I was able to get this to work. In the end, it was exactly as you wrote:

> I am not sure how you did it, but the JSON gRPC gateway in etcd seems to be disabled.
> You have to enable it back.

This specific requirement for etcd v3 is not detailed anywhere in the Patroni docs, for example:
https://patroni.readthedocs.io/en/latest/SETTINGS.html#etcdv3
> 
> Etcdv3
> If you want that Patroni works with Etcd cluster via protocol version 3, you need to use the etcd3 section in the Patroni configuration file. All configuration parameters are the same as for etcd.
> 
> Warning
> 
> Keys created with protocol version 2 are not visible with protocol version 3 and the other way around, therefore it is not possible to switch from etcd to etcd3 just by updating Patroni config file.
> 
A search for gRPC comes up empty. Interestingly enough, the etcd docs make no mention of the parameter needed. Which is:

```
# per https://github.com/etcd-io/etcd/issues/12093
enable-grpc-gateway: true
```

**Comment by ahrtr** [comment]
Is this project still using etcd's RESTful API (e.g. `http://127.0.0.1:2379/v3/cluster/member/list`)? Note that the etcd community is discussing to deprecate the RESTful API. 

FYI. https://github.com/etcd-io/etcd/pull/16049#issuecomment-1635983888

**Comment by TeoLefebvre** [comment]
I would like to comment that now, etcd grpc gateway is enabled by default even when etcd is launched with config file. I can't find when the changed this and the changelog but I tested it.
